### PR TITLE
add redisChannelPrefix option

### PR DESCRIPTION
**Pull request by kongkx**
Try to fix redis channel subscribe.

https://github.com/tlaverdure/laravel-echo-server/issues/400

### CODE REVIEW

File: src/cli/cli.ts [hunk 1]
```diff
@@ -184,10 +184,15 @@ export class Cli {
...
                 }
+            }, {
+                name: 'redisChannelPrefix',
+                default: '',
+                message: 'Enter the redis database prefix. if laravel 5.8, use `laravel_database_`, else leave it empty.',
             }, {
                 name: 'file',
                 default: defaultFile,
                 message: 'What do you want this config to be saved as?'
+
```
Comment (thomaswardiii) [inline src/cli/cli.ts]: Quibble: Remove the new line

File: src/subscribers/redis-subscriber.ts [hunk 2]
```diff
@@ -36,7 +36,9 @@ export class RedisSubscriber implements Subscriber {
                         Log.info("Event: " + message.event);
                     }
 
-                    callback(channel, message);
+                    const internalChannel = this.options.redisChannelPrefix ? channel.replace(new RegExp(`^${this.options.redisChannelPrefix}`), '') : channel;
```
Comment (thomaswardiii) [inline src/subscribers/redis-subscriber.ts]: I like that this is simple by not touching a lot of code, but is using regexp performant enough for potentially higher traffic sites?

Comment (kongkx) [reply src/subscribers/redis-subscriber.ts]: I'm sorry. I have not yet consider the performance impact. In fact, I'm wondering how can I do a test about this. If you got some idea,  let me know. I would like to do a test.

Comment (KristianI) [reply src/subscribers/redis-subscriber.ts]: I tried doing a simple test with a loop for the line in question:

```javascript
var start = Date.now();

var redisChannelPrefix = 'laravel_database_';
var channel = 'laravel_database_private-user.1234';
var i;

for (i = 0; i < 1000000; i++) {
  var result = redisChannelPrefix ? channel.replace(new RegExp(`^${redisChannelPrefix}`), '') : channel;
}

console.log(Date.now() - start);
```

It shows that it takes around 300 milliseconds (0.3 seconds) to run the RegExp line 1.000.000 times.

In my book that is fine.

Comment (thomaswardiii) [reply src/subscribers/redis-subscriber.ts]: Cool! I like this, but it's obviously up to the maintainer then

File: src/cli/cli.ts (hunk 1 above)
Comment (kongkx) [reply src/cli/cli.ts]: I have removed this new line.

**Comment by tthdvd** [comment]
When it will be released? It's quite important bug fix for me.

Thanks for the response!

**Comment by tlaverdure** [comment]
Could we just leverage the keyPrefix in ioredis?

https://github.com/luin/ioredis#transparent-key-prefixing

Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.

```
{
  databaseConfig: {
     keyPrefix: "laravel_database_"
  }
}
```

Thanks

**Comment by tthdvd** [comment]
> Could we just leverage the keyPrefix in ioredis?
> 
> https://github.com/luin/ioredis#transparent-key-prefixing
> 
> Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> 
> ```
> {
>   databaseConfig: {
>      keyPrefix: "laravel_database_"
>   }
> }
> ```
> 
> Thanks

Still does not work for me.
The config:
```
        "database": "redis",
        "databaseConfig": {
                "redis": {
                             "keyPrefix": "laravel_database_"
                        },
        },
```


And still can't get the message

[9:30:40 PM] - QPbaisz8B6JlVh7eAAAB authenticated for: private-App.User.1
[9:30:40 PM] - QPbaisz8B6JlVh7eAAAB joined channel: private-App.User.1
Channel: laravel_database_private-App.User.1
Event: App\Events\AcceptedBidEvent

**Comment by thomaswardiii** [comment]
> Could we just leverage the keyPrefix in ioredis?
> 
> https://github.com/luin/ioredis#transparent-key-prefixing
> 
> Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> 
> ```
> {
>    "databaseConfig": {
>      "redis": {
>        "keyPrefix": "laravel_database_"
>       }
>   }
> }
> ```
> 
> Thanks



> > Could we just leverage the keyPrefix in ioredis?
> > https://github.com/luin/ioredis#transparent-key-prefixing
> > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > ```
> > {
> >   databaseConfig: {
> >      keyPrefix: "laravel_database_"
> >   }
> > }
> > ```
> > 
> > 
> > Thanks
> 
> Still does not work for me.
> The config:
> 
> ```
>         "database": "redis",
>         "databaseConfig": {
>                 "redis": {
>                              "keyPrefix": "laravel_database_"
>                         }
>         },
> ```
> 
> And still can't listen the channel
> 
> [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB authenticated for: private-App.User.1
> [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB joined channel: private-App.User.1
> Channel: laravel_database_private-App.User.1
> Event: App\Events\AcceptedBidEvent

Weird. Looks like it should maybe work

**Comment by tthdvd** [comment]
> > Could we just leverage the keyPrefix in ioredis?
> > https://github.com/luin/ioredis#transparent-key-prefixing
> > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > ```
> > {
> >    "databaseConfig": {
> >      "redis": {
> >        "keyPrefix": "laravel_database_"
> >       }
> >   }
> > }
> > ```
> > 
> > 
> > Thanks
> 
> > > Could we just leverage the keyPrefix in ioredis?
> > > https://github.com/luin/ioredis#transparent-key-prefixing
> > > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > > ```
> > > {
> > >   databaseConfig: {
> > >      keyPrefix: "laravel_database_"
> > >   }
> > > }
> > > ```
> > > 
> > > 
> > > Thanks
> > 
> > 
> > Still does not work for me.
> > The config:
> > ```
> >         "database": "redis",
> >         "databaseConfig": {
> >                 "redis": {
> >                              "keyPrefix": "laravel_database_"
> >                         }
> >         },
> > ```
> > 
> > 
> > And still can't listen the channel
> > [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB authenticated for: private-App.User.1
> > [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB joined channel: private-App.User.1
> > Channel: laravel_database_private-App.User.1
> > Event: App\Events\AcceptedBidEvent
> 
> Weird. Looks like it should maybe work

I don't get it why this could solve the problem. 

The problem is that the Laravel framework give prefixes for the key what is stored in redis. The resolution could be this pull request, to match (with regex) and listen for the channels

**Comment by thomaswardiii** [comment]
> > > Could we just leverage the keyPrefix in ioredis?
> > > https://github.com/luin/ioredis#transparent-key-prefixing
> > > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > > ```
> > > {
> > >    "databaseConfig": {
> > >      "redis": {
> > >        "keyPrefix": "laravel_database_"
> > >       }
> > >   }
> > > }
> > > ```
> > > 
> > > 
> > > Thanks
> > 
> > 
> > > > Could we just leverage the keyPrefix in ioredis?
> > > > https://github.com/luin/ioredis#transparent-key-prefixing
> > > > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > > > ```
> > > > {
> > > >   databaseConfig: {
> > > >      keyPrefix: "laravel_database_"
> > > >   }
> > > > }
> > > > ```
> > > > 
> > > > 
> > > > Thanks
> > > 
> > > 
> > > Still does not work for me.
> > > The config:
> > > ```
> > >         "database": "redis",
> > >         "databaseConfig": {
> > >                 "redis": {
> > >                              "keyPrefix": "laravel_database_"
> > >                         }
> > >         },
> > > ```
> > > 
> > > 
> > > And still can't listen the channel
> > > [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB authenticated for: private-App.User.1
> > > [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB joined channel: private-App.User.1
> > > Channel: laravel_database_private-App.User.1
> > > Event: App\Events\AcceptedBidEvent
> > 
> > 
> > Weird. Looks like it should maybe work
> 
> I don't get it why this could solve the problem.
> 
> The problem is that the Laravel framework give prefixes for the key what is stored in redis. The resolution could be this pull request, to match (with regex) and listen for the channels

According to the documentation, the suggested fix tells the library being utilized (ioredis) to prefix that string infront of any string passed into the various key-related functions, so instead of putting the logic in the laravel-echo-server, we're attempting to use something built-in to the redis library it's using.

**Comment by tthdvd** [comment]
> > > > Could we just leverage the keyPrefix in ioredis?
> > > > https://github.com/luin/ioredis#transparent-key-prefixing
> > > > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > > > ```
> > > > {
> > > >    "databaseConfig": {
> > > >      "redis": {
> > > >        "keyPrefix": "laravel_database_"
> > > >       }
> > > >   }
> > > > }
> > > > ```
> > > > 
> > > > 
> > > > Thanks
> > > 
> > > 
> > > > > Could we just leverage the keyPrefix in ioredis?
> > > > > https://github.com/luin/ioredis#transparent-key-prefixing
> > > > > Can someone test adding this to their laravel-echo-server.json file to see if it fixes it.
> > > > > ```
> > > > > {
> > > > >   databaseConfig: {
> > > > >      keyPrefix: "laravel_database_"
> > > > >   }
> > > > > }
> > > > > ```
> > > > > 
> > > > > 
> > > > > Thanks
> > > > 
> > > > 
> > > > Still does not work for me.
> > > > The config:
> > > > ```
> > > >         "database": "redis",
> > > >         "databaseConfig": {
> > > >                 "redis": {
> > > >                              "keyPrefix": "laravel_database_"
> > > >                         }
> > > >         },
> > > > ```
> > > > 
> > > > 
> > > > And still can't listen the channel
> > > > [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB authenticated for: private-App.User.1
> > > > [9:30:40 PM] - QPbaisz8B6JlVh7eAAAB joined channel: private-App.User.1
> > > > Channel: laravel_database_private-App.User.1
> > > > Event: App\Events\AcceptedBidEvent
> > > 
> > > 
> > > Weird. Looks like it should maybe work
> > 
> > 
> > I don't get it why this could solve the problem.
> > The problem is that the Laravel framework give prefixes for the key what is stored in redis. The resolution could be this pull request, to match (with regex) and listen for the channels
> 
> According to the documentation, the suggested fix tells the library being utilized (ioredis) to prefix that string infront of any string passed into the various key-related functions, so instead of putting the logic in the laravel-echo-server, we're attempting to use something built-in to the redis library it's using.

I see, thanks for the answer.
unfortunately it doesn't work for me :(.

**Comment by tlaverdure** [comment]
Ok, I see now.

Laravel doesn't add the prefix to channels but it does to event broadcasting. I normally just turn redis prefixing off in my apps, so I haven't experienced this.

I think adding a new config option would add more confusion in getting setup. Let me think about this...

**Comment by KristianI** [comment]
> Ok, I see now.
> 
> Laravel doesn't add the prefix to channels but it does to event broadcasting. I normally just turn redis prefixing off in my apps, so I haven't experienced this.
> 
> I think adding a new config option would add more confusion in getting setup. Let me think about this...

These two comments explains why using keyPrefix from ioredis will not work solely:
https://github.com/tlaverdure/laravel-echo-server/issues/400#issuecomment-494348655
https://github.com/tlaverdure/laravel-echo-server/issues/400#issuecomment-516481202

This package doesn't work with the default Laravel setup as it is currently. Implementing this PR will correct that.

Recommending removal of redis prefix from Laravel to make this package works feels kind of weird, as the Laravel core team decided that the prefix was needed with the following explanation:
> This PR prefixes the redis database connection by default to mitigate multiple sites on the same server potentially sharing the same queued jobs. It uses similar logic to that which is used for creating the cache prefix.

Hope we can reach a good solution on this 🙏

**Comment by kongkx** [comment]
I think there should be more code to be added for handling "multiple sites on the same server potentially sharing the same queued jobs".  
- laravel app "App_A" with redis prefix `A_database_`
- laravel app "App_B" with redix prefix `B_database_`

- laravel-echo-server A for AppA will also trying to handle `B_database_private.User.1` message from `App_B`, although no message will be broadcasted through "socket-io" server.

Maybe we should test channel before calling the callback in `RedisSubscriber.subscribe`. 

```js
_this._redis.on('pmessage', function(subscribed, channel, message) {
    try {
        message = JSON.parse(message)
        if (_this.options.devMode) {
            log_1.Log.info('Channel: ' + channel)
            log_1.Log.info('Event: ' + message.event)
        }
        // appRedisPrefix may be a better config name
        var internalChannel
        if (this.options.appRedisPrefix) {
            var prefixRegex = new RegExp(
                '^' + _this.options.appRedisPrefix
            )
            internalChannel = channel.replace(prefixRegex, '')
            if (channel === internalChannel) {
                log_1.Log.info('Not target app event, not handled')
                return
            }
        } else {
            internalChannel = channel
        }
        callback(internalChannel, message)
    } catch (e) {
        if (_this.options.devMode) {
            log_1.Log.info('No JSON message')
        }
    }
})
```

**Comment by tlaverdure** [comment]
I think using `keyPrefix` should just work.
 
I sent a PR to laravel/framework to ignore the prefix when verifying channel authorization in `channels.php`.

https://github.com/laravel/framework/pull/30597

**Comment by kongkx** [comment]
> I think using `keyPrefix` should just work.
> 
> I sent a PR to laravel/framework to ignore the prefix when verifying channel authorization in `channels.php`.
> 
> [laravel/framework#30597](https://github.com/laravel/framework/pull/30597)

With your pull request for `laravel/framework`, `keyPrefix` should work.

But I think it's quite confusing about the usage of `keyPrefix` (`databaseConfig.redis`), as it's used both in data storage and subscriber.  In my opinion, it needs to be separated.  

- data may conflict between laravel and laravel-echo-server in redis, although some cacke key pattern may help to solve this problem.
-  laravel-echo-server may using another redis service for data storage which is not for broadcasting

**Comment by litan1106** [comment]
I ran into this issue as well.

**Comment by tlaverdure** [comment]
Closed by #464. Please also upgrade Laravel Framework to 6.5.2

**Pull request closed without merging**